### PR TITLE
add generate backfill table helper

### PR DESCRIPTION
**Pull request by ShakoHo**
@askeing @MikeLien @ypwalter r?

### CODE REVIEW

File: ejenti/tasks/hasalTasks.py [hunk 1]
```diff
@@ -66,14 +67,77 @@ def checkout_latest_code(**kwargs):
     git_pull(**kwargs)
 
 
+def backfill_hasal_on_perfherder_data(**kwargs):
```
Comment (askeing) [inline ejenti/tasks/hasalTasks.py]: base on the `run_hasal_on_latest_nightly()`, should we rename the method name from `backfill_hasal_on_perfherder_data()` to `run_hasal_on_specify_nightly()`?

File: requirements.txt [hunk 2]
```diff
@@ -14,8 +14,10 @@ selenium==3.0.2
 jsonschema==2.6.0
 geckoprofiler_controller==0.0.4
 requests==2.18.4
-
+lxml==4.0.0
 psutil==5.3.0
+beautifulsoup4==4.6.0
+html5lib>=0.999
```
Comment (askeing) [inline requirements.txt]: should be `==`

File: lib/common/commonUtil.py [hunk 3]
```diff
@@ -560,6 +562,104 @@ def deep_merge_dict(source_dict, destination_dict):
...
+                for data in tqdm(response.iter_content(chunk_size=512 * 1024), total=total_len / (512 * 1024)):
+                    fh.write(data)
+            return download_fp
+        except Exception as e:
+            logger.error("ERROR: [%s]" % e.message)
+            return None
+
+    @staticmethod
+    def get_files_from_remote_url_folder(input_url, input_params=None):
+        return_dict = {}
```
Comment (askeing) [inline lib/common/commonUtil.py]: adding the comment here to describe the return object will be better.

(the name `get_files_...` looks like return a array.)

File: lib/common/commonUtil.py [hunk 4]
```diff
@@ -560,6 +562,104 @@ def deep_merge_dict(source_dict, destination_dict):
...
+        """
+        try:
+            if input_data and input_headers:
+                response_obj = requests.post(input_url, data=input_data, headers=input_headers)
+            elif input_data:
+                response_obj = requests.post(input_url, data=input_data)
+            elif input_headers:
+                response_obj = requests.post(input_url, headers=input_headers)
+            else:
+                response_obj = requests.post(input_url)
```
Comment (askeing) [inline lib/common/commonUtil.py]: due to the default value of `input_data` and `input_headers` are `None`, it seems like we can just call `response_obj = requests.post(input_url, data=input_data, headers=input_headers)`.

File: lib/common/commonUtil.py [hunk 5]
```diff
@@ -560,6 +562,104 @@ def deep_merge_dict(source_dict, destination_dict):
...
+        """
+        try:
+            if input_params and input_headers:
+                response_obj = requests.get(input_url, params=input_params, headers=input_headers)
+            elif input_params:
+                response_obj = requests.get(input_url, params=input_params)
+            elif input_headers:
+                response_obj = requests.get(input_url, headers=input_headers)
+            else:
+                response_obj = requests.get(input_url)
```
Comment (askeing) [inline lib/common/commonUtil.py]: due to the default value of input_data and input_headers are None, it seems like we can just call response_obj = requests. get(input_url, data=input_data, headers=input_headers).

File: lib/helper/archiveMozillaHelper.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            response_table_dict = response_df.set_index("Name").T.to_dict()
+            folder_name_list = [d_name for d_name in response_table_dict.keys() if d_name != ".."]
+            folder_name_list.sort()
+            return folder_name_list
+        else:
+            return []
+
+    @staticmethod
+    def filter_backfill_period_data(input_backfill_start_date, input_current_date, input_backfill_repo, input_folder_list, input_query_channel_archive_url):
+        result_dict = {}
```
Comment (askeing) [inline lib/helper/archiveMozillaHelper.py]: need comment for this method's return value.

File: lib/helper/archiveMozillaHelper.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        DEFAULT_QUERY_CHANNEL_ARCHIVE_URL = "%s/pub/%s/%s" % (ArchiveMozillaHelper.DEFAULT_ARCHIVE_URL, input_app_name, input_channel_name)
+        query_dir_url = "%s/%d/%02d/" % (DEFAULT_QUERY_CHANNEL_ARCHIVE_URL, int(backfill_start_year), int(backfill_start_month))
+        total_folder_list = ArchiveMozillaHelper.get_folder_name_list(query_dir_url)
+        if backfill_start_year != current_year or backfill_start_month != current_month:
+            query_dir_url = "%s/%d/%02d/" % (DEFAULT_QUERY_CHANNEL_ARCHIVE_URL, int(current_year), int(current_month))
+            total_folder_list.extend(ArchiveMozillaHelper.get_folder_name_list(query_dir_url))
+        for history_folder_name in input_history_folder_list:
+            if history_folder_name not in total_folder_list:
+                total_folder_list.append(history_folder_name)
+        return ArchiveMozillaHelper.filter_backfill_period_data(backfill_start_date_without_time, current_utc_time, input_backfill_repo, total_folder_list, DEFAULT_QUERY_CHANNEL_ARCHIVE_URL)
```
Comment (askeing) [inline lib/helper/archiveMozillaHelper.py]: The `filter_backfill_period_data()` will return the dict object, but the method name is `get...list()`.
Also need some comments to describe the `input_app_name`, `input_channel_name`, and `input_history_folder_list`, for example, `https://archive.mozilla.org/pub/<input_app_name>/<input_channel_name>`

File: lib/helper/generateBackfillTableHelper.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+logger = get_logger(__name__)
+
+
+class GenerateBackfillTableHelper(object):
+    DEFAULT_HG_QUERY_REVISION_JSON_URL = "https://hg.mozilla.org/mozilla-central/json-rev"
+    DEFAULT_BACKFILL_TABLE_LOCAL_FN = "backfill_table.json"
+
+    @staticmethod
+    def generateArchieveRevisionRelationTable(input_history_backfill_table, input_backfill_days=14, input_backfill_repo="mozilla-central", input_app_name="firefox", input_channel_name="nightly"):
```
Comment (askeing) [inline lib/helper/generateBackfillTableHelper.py]: modify the method name to `abc_xyz_foo_bar` will better

File: lib/helper/generateBackfillTableHelper.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                    "pkg_json_url": history_backfill_url_dict[current_backfill_dir_url]["pkg_json_url"],
+                    "pkg_fn_url": history_backfill_url_dict[current_backfill_dir_url]["pkg_fn_url"],
+                    "archive_url": current_backfill_dir_url,
+                    "archive_dir": history_backfill_url_dict[current_backfill_dir_url]["archive_dir"],
+                    "archive_datetime": history_backfill_url_dict[current_backfill_dir_url]["archive_datetime"]}
+
+        return archieve_revision_relation_table
+
+    @staticmethod
+    def generateArchievePerfherderReleationTable(input_backfill_days=14, input_backfill_repo="mozilla-central", input_app_name="firefox", input_channel_name="nightly", input_white_list=[]):
```
Comment (askeing) [inline lib/helper/generateBackfillTableHelper.py]: modify the method name to `abc_xyz_foo_bar` will better

File: lib/helper/perfherderDataQueryHelper.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                    else:
+                        input_result_dict[push_tiemstamp]['perfherder_data'][perfherder_data_index] = {"value": [data['value']],
+                                                                                                       "signature": sig}
+                else:
+                    input_result_dict[push_tiemstamp] = {'perfherder_data': {perfherder_data_index: {"value": [data['value']],
+                                                                                                     "signature": sig}}}
+        return input_result_dict
+
+    @staticmethod
+    def get_perfherder_data(input_white_list, input_query_days=14, input_query_channel="mozilla-central", input_hasal_framework_no=9):
```
Comment (askeing) [inline lib/helper/perfherderDataQueryHelper.py]: we need the comments to describe the input parameters, like the format of white list.

File: lib/common/commonUtil.py (hunk 4 above)
Comment (ShakoHo) [reply lib/common/commonUtil.py]: without giving parameters behave differently with specifying the parameter with None value

File: lib/common/commonUtil.py (hunk 5 above)
Comment (ShakoHo) [reply lib/common/commonUtil.py]: without giving parameters behave differently with specifying the parameter with None value

File: lib/common/commonUtil.py (hunk 4 above)
Comment (askeing) [reply lib/common/commonUtil.py]: wow, it's really mysterious.
base on the code [api.py](https://github.com/requests/requests/blob/v2.18.4/requests/api.py#L16-L58) and [sessions.py](https://github.com/requests/requests/blob/v2.18.4/requests/sessions.py#L441-L444), the default value should be None 🤔

File: lib/helper/generateBackfillTableHelper.py (hunk 8 above)
Comment (ShakoHo) [reply lib/helper/generateBackfillTableHelper.py]: this is the official method for another python to called, so should be named it more meaningful

File: lib/helper/generateBackfillTableHelper.py (hunk 9 above)
Comment (ShakoHo) [reply lib/helper/generateBackfillTableHelper.py]: this is the official method for another python to called, so should be named it more meaningful

File: lib/helper/perfherderDataQueryHelper.py (hunk 10 above)
Comment (ShakoHo) [reply lib/helper/perfherderDataQueryHelper.py]: already updated

File: lib/helper/generateBackfillTableHelper.py (hunk 8 above)
Comment (askeing) [reply lib/helper/generateBackfillTableHelper.py]: `generate_archieve_revision_relation_table()`

File: lib/helper/generateBackfillTableHelper.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                    "pkg_json_url": history_backfill_url_dict[current_backfill_dir_url]["pkg_json_url"],
+                    "pkg_fn_url": history_backfill_url_dict[current_backfill_dir_url]["pkg_fn_url"],
+                    "archive_url": current_backfill_dir_url,
+                    "archive_dir": history_backfill_url_dict[current_backfill_dir_url]["archive_dir"],
+                    "archive_datetime": history_backfill_url_dict[current_backfill_dir_url]["archive_datetime"]}
+
+        return archieve_revision_relation_table
+
+    @staticmethod
+    def generate_archieve_perfherder_releation_table(input_backfill_days=14, input_backfill_repo="mozilla-central", input_app_name="firefox", input_channel_name="nightly", input_white_list=[]):
```
Comment (askeing) [inline lib/helper/generateBackfillTableHelper.py]: After having a discussion, it seems like the main entry point.
IMO, this method needs comments to tell developer `I am entry point ^.<y` and parameters and return values.

File: lib/helper/generateBackfillTableHelper.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+                     'revision': u'e4261f5b96ebfd63e7cb8af3035ff9fea90c74a5'}
+        }
+        """
+        archieve_revision_relation_table = {}
+        backfill_repo_name = input_backfill_repo + "/"
+        history_backfill_url_dict = {}
+        history_backfill_dir_list = []
+
+        # init platform value
+        if sys.platform == "linux2":
```
Comment (askeing) [inline lib/helper/generateBackfillTableHelper.py]: should adding `input_platform` parameter for selecting specify platform, and only using current platform when input parameter is None

**Pull request merged**